timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/LeadershipGuardedClientAwareManagedTimestampService.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.transaction.timestamp;
+
+import com.palantir.leader.NotCurrentLeaderException;
+import com.palantir.lock.v2.PartitionedTimestamps;
+import com.palantir.timestamp.TimestampRange;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link ClientAwareManagedTimestampService} that also verifies that it is not known to not be the leader before
+ * returning results to the user.
+ */
+public class LeadershipGuardedClientAwareManagedTimestampService
```

Comment:
`@AutoDelegate`? 🔥

Reply:
I think we'd have to override everything anyway 😅